lib/file-source/src/file_watcher.rs

```diff
@@ -61,28 +61,57 @@ impl FileWatcher {
             false
         };
 
-        let (reader, file_position): (Box<dyn BufRead>, FilePosition) = if is_gzipped(&mut reader)?
-        {
-            if file_position != 0 || too_old {
-                // We can't accurately seek into gzipped files without manually scanning through
-                // the entire thing, so for now we simply refuse to read gzipped files for which we
-                // already have a stored file position from a previous run.
-                debug!(
-                    message = "Not re-reading gzipped file with existing stored offset.",
-                    ?path,
-                    %file_position
-                );
-                (Box::new(null_reader()), file_position)
-            } else {
-                (Box::new(io::BufReader::new(MultiGzDecoder::new(reader))), 0)
-            }
-        } else if too_old {
-            let pos = reader.seek(io::SeekFrom::End(0)).unwrap();
-            (Box::new(reader), pos)
-        } else {
-            let pos = reader.seek(io::SeekFrom::Start(file_position)).unwrap();
-            (Box::new(reader), pos)
-        };
+        let gzipped = is_gzipped(&mut reader)?;
+
+        // Determine the actual position at which we should start reading
+        let (reader, file_position): (Box<dyn BufRead>, FilePosition) =
+            match (gzipped, too_old, read_from) {
+                (true, true, _) => {
+                    debug!(
+                        message = "Not reading gzipped file older than `ignore_older`.",
+                        ?path,
+                    );
+                    (Box::new(null_reader()), 0)
+                }
+                (true, _, ReadFrom::Checkpoint(file_position)) => {
+                    debug!(
+                        message = "Not re-reading gzipped file with existing stored offset.",
+                        ?path,
+                        %file_position
+                    );
+                    (Box::new(null_reader()), file_position)
+                }
+                // TODO: This may become the default, leading us to stop reading gzipped files that
+                // we were reading before. Should we merge this and the next branch to read
+                // compressed file from the beginning even when `read_from = "end"` (implicitly via
+                // default or explicitly via config)?
+                (true, _, ReadFrom::End) => {
+                    debug!(
+                        message = "Can't read from the end of already-compressed file.",
+                        ?path,
+                    );
+                    (Box::new(null_reader()), 0)
+                }
```

Comment:
This is probably the most interesting open question here. We basically need to decide how `read_from` applies to compressed files.

Right now it's not as big of a deal because we default to reading from the beginning, which works with compressed files just fine. But it does mean that if users specify `read_from = "end"` then we will not ingest compressed files. This seems to make sense, since `end` implies that we only want to tail new files, which doesn't really apply to compressed files.

However, when we change the default from `beginning` to `end`, this becomes trickier. If we don't adjust this at all, then we would stop ingesting compressed files by default, which seems potentially surprising. Users would need to opt in to `read_from = "beginning"` to get compressed files. We could also choose to ignore `read_from` for compressed files and assume that if users specify them in `include` that they want to ingest them, even if it doesn't match up with `read_from`.

/cc @binarylogic

Reply:
Both options have something going for them so I agree it would better to move this decision to it's own PR where it will get a proper attention as it would be a breaking change.

Reply:
What's the problem with reading the gzipped files from the end/checkpoint?

Reply:
I presume the reads from the end of the not-seen-before files are a real problem (well, we could always read-through the files till the end) - but for checkpoints, we could possibly keep some context around to allow resuming from where we left off.

Reply:
We can't really seek into compressed files. We could hack around it by reading and discarding but it'd be inefficient and I'd rather not unless we have a very good reason. Also, in most log rotation cases I can think of, compression is something that happens once a file is done being written to. It should be rare that someone wants to "tail" a compressed file, and I don't think we should spend the time to implement it without some clear demand.

Reply:
> It should be rare that someone wants to "tail" a compressed file, and I don't think we should spend the time to implement it without some clear demand.

I agree, but the case for reading the compressed files from the middle can be created by Vector itself - if we read through the compressed file half-way, and then shutdown, we can't continue from where we left off. As such, the unread portion of the file will be forever lost.
From this perspective, maybe it is better to read and discard, rather than just dropping half of the data?

Waiting for the user demand on this is like waiting for our users to test our code and find the bug we've already known about. If the user demand is not clear - maybe we can drop support for the compressed files altogether? Occasionally dropping half of the files might be worse than not reading them at all.

Reply:
Added #6284 to track this.